### PR TITLE
[AutoParallel] Support sharding strategy to_static

### PR DESCRIPTION
**Pull request by LiYuRio**
### PR types

New features

### PR changes

Others

### Description

通过配置strategy的方式，支持sharding策略的动转静

PCard-73145

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/api.py [hunk 1]
```diff
@@ -604,14 +598,56 @@ def __init__(self, optimizer, shard_fn=None):
...
+                if self._sharding_degree is None:
+                    # set the sharding degree if it has not been set
+                    if any(
+                        isinstance(placement, dist.Shard)
+                        for placement in placements
+                    ):
+                        for idx, placement in enumerate(placements):
+                            if isinstance(placement, dist.Replicate):
+                                self._sharding_degree = mesh.dim_size(idx)
+                                self._sharding_mesh_axis = idx
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py:634]: break?

File: python/paddle/distributed/auto_parallel/api.py [hunk 2]
```diff
@@ -604,14 +598,56 @@ def __init__(self, optimizer, shard_fn=None):
...
+                else:
+                    # check the sharding degree since it has already been set
+                    if any(
+                        isinstance(placement, dist.Shard)
+                        for placement in placements
+                    ):
+                        for idx, placement in enumerate(placements):
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_parallel/api.py:646]: check placement[self._sharding_mesh_axis] is replicated?

File: python/paddle/distributed/auto_parallel/api.py (hunk 1 above)
Comment (LiYuRio) [reply python/paddle/distributed/auto_parallel/api.py:634]: done

File: python/paddle/distributed/auto_parallel/api.py (hunk 2 above)
Comment (LiYuRio) [reply python/paddle/distributed/auto_parallel/api.py:646]: done

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**